src/buffer/out/textBuffer.cpp

```diff
@@ -563,14 +563,41 @@ til::point TextBuffer::NavigateCursor(til::point position, til::CoordType distan
 
 // This function is intended for writing regular "lines" of text as it'll set the wrap flag on the given row.
 // You can continue calling the function on the same row as long as state.columnEnd < state.columnLimit.
-void TextBuffer::Write(til::CoordType row, const TextAttribute& attributes, RowWriteState& state)
+void TextBuffer::Replace(til::CoordType row, const TextAttribute& attributes, RowWriteState& state)
 {
     auto& r = GetMutableRowByOffset(row);
     r.ReplaceText(state);
     r.ReplaceAttributes(state.columnBegin, state.columnEnd, attributes);
     TriggerRedraw(Viewport::FromExclusive({ state.columnBeginDirty, row, state.columnEndDirty, row + 1 }));
 }
 
+void TextBuffer::Insert(til::CoordType row, const TextAttribute& attributes, RowWriteState& state)
+{
+    auto& r = GetMutableRowByOffset(row);
+    auto& scratch = GetScratchpadRow();
+
+    scratch.CopyFrom(r);
+
+    r.ReplaceText(state);
+    r.ReplaceAttributes(state.columnBegin, state.columnEnd, attributes);
+
+    // Restore trailing text from our backup in scratch.
+    RowWriteState restoreState{
+        .text = scratch.GetText(state.columnBegin, state.columnLimit),
+        .columnBegin = state.columnEnd,
+        .columnLimit = state.columnLimit,
+    };
+    r.ReplaceText(restoreState);
+
+    // Restore trailing attributes as well.
+    auto& rowAttr = r.Attributes();
+    const auto& scratchAttr = scratch.Attributes();
+    const auto restoreAttr = scratchAttr.slice(gsl::narrow<uint16_t>(state.columnBegin), gsl::narrow<uint16_t>(state.columnLimit));
+    rowAttr.replace(gsl::narrow<uint16_t>(state.columnEnd), gsl::narrow<uint16_t>(state.columnEnd + restoreAttr.size()), restoreAttr);

```

Comment:
@DHowett Ah no Dustin, this isn't quite correct. You see...

If you had a 6 column row and the contents were
```
woooo
012356
```

and you add "lol" at 2, like
```
wololo
0123456
```

![image](https://github.com/microsoft/terminal/assets/2256941/6b0f95bf-f49a-42c1-b25a-38c16e770451)

it should copy attributes `[5,6)` ("o"), but I pass I pass `[columnBegin,columnLimit)` which is `[2,6)` ("ooo"). Before I tried adding support for margins this worked, because I finished it up with a `resize_trailing_extent` which fixed this up, but now we can't do this anymore.

What I should do is copy `restoreState.columnBegin,restoreState.columnEnd)` which should be immediately obvious because that's literally the exact same thing we pass to `ReplaceAttributes` above!

I'll write tests, as I should've always done...

Reply:
Ha thanks for confessing before it merged! ;)